### PR TITLE
Make grid configurable from text file

### PR DESCRIPTION
**Pull request by marfvr**
## Proposed changes

Make grid configurable from a text file.

## Fixes

n/a

## Types of changes

What types of changes does your code introduce?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply._

- [ ] I have read the [CONTRIBUTING](../blob/master/CONTRIBUTING.md) doc
- [ ] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [ ] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works

## Further comments

n/a

### CODE REVIEW

File: gym_sapientino/assets/default_map.txt [hunk 1]
```diff
@@ -0,0 +1,5 @@
+P bB g |
+ bp G r|
+G   pg |
+ rpG PB|
+rP Bg b|
```
Comment (cipollone) [inline gym_sapientino/assets/default_map.txt]: Very handy!
Is there a documentation about this format? (even a two lines docstring in some class)

File: gym_sapientino/core/configurations.py [hunk 2]
```diff
@@ -68,12 +71,35 @@ class SapientinoConfiguration:
     agent_configs: Tuple[SapientinoAgentConfiguration, ...] = (
         SapientinoAgentConfiguration(),
     )
-    rows: int = 5
-    columns: int = 7
+    path_to_map: Path = ASSETS_DIR / DEFAULT_MAP_FILENAME
```
Comment (cipollone) [inline gym_sapientino/core/configurations.py:74]: Actually, I didn't know this operator on paths. Nice

File: gym_sapientino/core/grid.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    def __init__(self, cells: List[List[Cell]]):
+        """Initialize the grid."""
+        self.cells: List[List[Cell]] = cells
+        self.color_count: Dict[Colors, int] = {}
+
+    @property
+    def rows(self):
+        """Get the number of rows."""
+        # TODO allow different number of rows and columns.
+        return len(self.cells)
```
Comment (cipollone) [inline gym_sapientino/core/grid.py:72]: What do you mean by "different number of rows and columns"?
Non rectangular map? That yould be also nice, and not too hard either, but I don't know if the transition functions assume a square shape somewhere.

**Review by cipollone — APPROVED** [review]
Didn't test it, but looks good!

File: gym_sapientino/core/grid.py (hunk 3 above)
Comment (marfvr) [reply gym_sapientino/core/grid.py:72]: no, Sapientino doesn't assume a square shape. You're right, it's not that hard, but maybe for our purposes, this is good enough. We will see

File: gym_sapientino/assets/default_map.txt (hunk 1 above)
Comment (marfvr) [reply gym_sapientino/assets/default_map.txt]: You're right, we need documentation. I'll add some comments somewhere

File: gym_sapientino/core/grid.py (hunk 3 above)
Comment (cipollone) [reply gym_sapientino/core/grid.py:72]: It is! we don't need it now. So rows and columns are correct, right? Why the TODOs.
Also, in the parsing function, you could do something like `all((len(cell_str[i]) == nb_columns for i in nb_rows))`
Nothing important btw

Comment (marfvr) [reply gym_sapientino/core/grid.py:72]: Will fix in `from_map` during the inner for loop

File: gym_sapientino/core/grid.py [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+    content = path_to_map.read_text()
+    cells_str = [list(line[:-1]) for line in content.splitlines(keepends=False)]
+    assert len(cells_str) > 0, "No row found."
+    assert len(cells_str[0]) > 0, "No column found."
+    nb_rows, nb_columns = len(cells_str), len(cells_str[0])
+    cells = []
+    for i in range(nb_rows):
+        row = []
+        for j in range(nb_columns):
+            assert len(cells_str[i]) == nb_columns, "Got rows of different size"
```
Comment (marfvr) [inline gym_sapientino/core/grid.py:133]: @cipollone

File: gym_sapientino/assets/default_map.txt (hunk 1 above)
Comment (marfvr) [reply gym_sapientino/assets/default_map.txt]: @cipollone 

https://github.com/whitemech/gym-sapientino/blob/e6f2bae06be06b19f23ae07fea0c2a5a37f06eef/docs/configure_map.ipynb

Comment (cipollone) [reply gym_sapientino/assets/default_map.txt]: very nice

**Pull request merged**